add SignKey gpg entity param to allow easier pgp signing of commits

anandkumarpatel opened the pull request:
Reopening of https://github.com/google/go-github/pull/1142

This PR creates a new param for the `Commit` struct called `SignKey`.
`SignKey` is an `openpgp.Entity` type which is the standard way of reading and validating pgp keys.

When specified the key will be used to sign the commit. This makes it so the user does not have to sign the key externally and duplicate some of the logic the `CreateCommit` function does.

To maintain backward compatibility, if the `Verification.Signature` is defined `SignKey` will have no effect. 

This provides similar functionality that go-git has: https://github.com/src-d/go-git/commit/7b6c1266556f59ac436fada3fa6106d4a84f9b56

Note from the git docs: https://github.com/octokit/routes/blob/152c0b5d156844e17a1805d0509c9a0b7cfc848f/routes/api.github.com/git.json#L282
```
To pass a `signature` parameter, you need to first manually create a valid PGP signature, which can be complicated. You may find it easier to [use the command line]
```
So I wanted to make it less complicated 😄

gmlewis commented on github/git_commits_test.go:
```diff
@@ -123,6 +127,172 @@ func TestGitService_CreateSignedCommit(t *testing.T) {
+		Date:  &date,
+	}
+	input := &Commit{
+		Message:    String("m"),
```
Can you please change `"m"` to `"Commit message."`, as it makes line 169 much easier to read below. Thanks.

gmlewis commented on github/git_commits_test.go:
```diff
@@ -123,6 +127,172 @@ func TestGitService_CreateSignedCommit(t *testing.T) {
+	if err == nil {
+		t.Errorf("Expected error to be returned because no author was passed")
+	}
+	expectedString := "Commit Author is required to sign a commit"
```
Let's please not test against the error message value.... you can remove this section completely.
That way we can easily change it. :smile:

gmlewis commented on github/git_commits_test.go:
```diff
@@ -123,6 +127,172 @@ func TestGitService_CreateSignedCommit(t *testing.T) {
+	date, _ := time.Parse("Mon Jan 02 15:04:05 2006 -0700", "Thu May 04 00:03:43 2017 +0200")
+
+	msg := createSignatureMessage(&createCommit{
+		Message: String("m"),
```
Same here, please: `"Commit message."`.

gmlewis commented on github/git_commits_test.go:
```diff
@@ -123,6 +127,172 @@ func TestGitService_CreateSignedCommit(t *testing.T) {
+	date, _ := time.Parse("Mon Jan 02 15:04:05 2006 -0700", "Thu May 04 00:03:43 2017 +0200")
+
+	msg := createSignatureMessage(&createCommit{
+		Message: String("m"),
```
And here.

gmlewis commented on github/git_commits_test.go:
```diff
@@ -131,3 +301,62 @@ func TestGitService_CreateCommit_invalidOwner(t *testing.T) {
 	testURLParseError(t, err)
 }
+
+const testKey = `
```
`testGPGKey` please.

anandkumarpatel commented on github/git_commits.go:
```diff
@@ -133,3 +148,41 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+	message := ""
+
+	if commit.Tree != nil {
+		message = fmt.Sprintf("tree %s\n", *commit.Tree)
```
There are no getters defined for `createCommit` object

gmlewis replied:
Yes, whups, my bad.  Thanks, @anandkumarpatel.

anandkumarpatel commented on github/git_commits.go:
```diff
@@ -133,3 +148,41 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+
+func createSignature(SigningKey *openpgp.Entity, commit *createCommit) (string, error) {
+	if commit.Author == nil {
+		return "", fmt.Errorf("Commit Author is required to sign a commit")
```
Strange.. go vet didn't find a problem for me. Changing the error message anyway, I like the searchability aspect.

anandkumarpatel commented on github/git_commits.go:
```diff
@@ -133,3 +148,41 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+		message += fmt.Sprintf("parent %s\n", parent)
+	}
+
+	message += fmt.Sprintf("author %s <%s> %d %s\n", *commit.Author.Name, *commit.Author.Email, commit.Author.Date.Unix(), commit.Author.Date.Format("-0700"))
```
Same no getters for `createCommit` object. Getters are only defined for the normal `Commit` object not the one created in this file for some reason.

anandkumarpatel commented on github/git_commits.go:
```diff
@@ -133,3 +148,41 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+	}
+
+	message += fmt.Sprintf("author %s <%s> %d %s\n", *commit.Author.Name, *commit.Author.Email, commit.Author.Date.Unix(), commit.Author.Date.Format("-0700"))
+	commiter := commit.Committer
```
nice catch!

anandkumarpatel commented:
@gmlewis fixed most of the comments you left. The only issue was with using getters on the `createCommit` object. they do not exist so I left those as is but used getters on the sub-objects that had them. 

I also updated all the short names in the test file and defined them more verbosely.

gmlewis commented on github/git_commits.go:
```diff
@@ -176,13 +176,14 @@ func createSignatureMessage(commit *createCommit) string {
+	committer := commit.Committer
+	if committer == nil {
+		committer = commit.Author
 	}
```
Ad this point, I still think we need a test if `committer==nil` since both `commit.Committer` and `commit.Author` could be `nil`.

anandkumarpatel replied:
I check if `commit.Author` is nil at the top of the function since this function cannot work if that does not exist. Nothing modifies commit object in this function so not sure if we need to test for `nil` again?

https://github.com/google/go-github/pull/1198/files#diff-7da0b51e99c908bfec3fe2f646d4a0fdR154

gmlewis replied:
Ah, I forgot about that.  Great, thanks!  You are correct... that is sufficient.  I'm trying to avoid crashes and that settles it. I'm going to add a tweak above, though.

gmlewis commented on github/git_commits.go:
```diff
@@ -176,13 +176,14 @@ func createSignatureMessage(commit *createCommit) string {
 	// There needs to be a double newline after committer
-	message += fmt.Sprintf("committer %s <%s> %d %s\n\n", *commiter.Name, *commiter.Email, commiter.Date.Unix(), commiter.Date.Format("-0700"))
+	message += fmt.Sprintf("committer %s <%s> %d %s\n\n", committer.GetName(), committer.GetEmail(), committer.GetDate().Unix(), committer.GetDate().Format("-0700"))
 	message += fmt.Sprintf("%s", *commit.Message)
```
We also need a test if `commit.Message` is `nil` if there is no getter for it.

anandkumarpatel replied:
added check to the caller function

gmlewis commented:
> I also updated all the short names in the test file and defined them more verbosely.

Just FYI... the other names were fine.  As you'll see in a lot of other tests, we use `o` for `org` and `r` for `repo` all over the place, and it would be nice to be consistent for these...

But is was just a bit challenging at a quick glance to figure out in the tests that:

```
...

m
```

was the commit message.  Thanks.

gmlewis commented on github/git_commits.go:
```diff
@@ -133,3 +149,41 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
 }
+
+func createSignature(SigningKey *openpgp.Entity, commit *createCommit) (string, error) {
+	if commit.Author == nil {
```
Let's please make this:

```go
func createSignature(signingKey *openpgp.Entity, commit *createCommit) (string, error) {
  if signingKey == nil || commit == nil || commit.Author == nil {
    return "", errors.New("createSignature: invalid parameters")
  }
```

gmlewis replied:
Note the lower-case `signingKey`... no need to capitalize that parameter.

anandkumarpatel commented:
@gmlewis ready for round 2 👊.

* added nil checks to both functions
* added respective tests for those nil checks
* renamed variable to start lower case

gmlewis commented on github/git_commits.go:
```diff
@@ -133,3 +149,49 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+
+	message, err := createSignatureMessage(commit)
+	if err != nil {
+		return "", fmt.Errorf("createSignature: createSignatureMessage error: %s", err.Error())
```
Let's please change this line to `return "", err` since the child error identifies itself.

gmlewis commented on github/git_commits.go:
```diff
@@ -133,3 +149,49 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+		return "", errors.New("createSignatureMessage: invalid parameters")
+	}
+
+	message := ""
```
I'm just going to throw this out for your consideration, and I'm fine either way.
When I'm building up potentially long strings that are a collection of lines, I often like to use a slice of strings and then join them at the end.  I think it is much easier this way to write and also to maintain... because eventually someone will come along and add another line but forget to add their `\n`... so it looks something like this:

```go
var lines []string
...
lines = append(lines, fmt.Sprintf("...", ...))  // no \n
...
return strings.Join(lines, "\n"), err
```

Totally your call.  I prefer it, but this way is fine too.

anandkumarpatel replied:
I like this, does make it easier to modify!

anandkumarpatel commented:
@gmlewis Just pushed, lmk if there is something else.

gmlewis commented on github/git_commits.go:
```diff
@@ -133,3 +150,51 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+	}
+
+	// There needs to be a double newline after committer
+	message = append(message, fmt.Sprintf("committer %s <%s> %d %s\n", committer.GetName(), committer.GetEmail(), committer.GetDate().Unix(), committer.GetDate().Format("-0700")))
```
On this line you can remove the `\n`.

anandkumarpatel replied:
Yea I was debating what to do with this one. We need 2 new lines after the committer for this the be valid. It was between this or doing an empty string 
```
	message = append(message, fmt.Sprintf(""))
```
Though not sure if that is better or not.

gmlewis replied:
Oh, sorry... yes, if you want a blank line definitely keep the `\n` here.  No, never use `fmt.Sprintf("")`... that would just be a `""` string literal... but in this case, please just keep the `\n`.

gmlewis commented on github/git_commits.go:
```diff
@@ -133,3 +150,51 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+}
+
+func createSignatureMessage(commit *createCommit) (string, error) {
+	if commit == nil || commit.Message == nil || commit.Author == nil {
```
It just dawned on me that `commit.Message == nil` should be `commit.GetMessage() == ""` because we also care that there is no message, and this helper function does the extra work for us.

I apologize for not catching this one before.

anandkumarpatel replied:
Remember there are no getters for type `createCommit`

But will check for `nil` and empty string.

gmlewis commented on github/git_commits.go:
```diff
@@ -133,3 +150,51 @@ func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string
+
+	// There needs to be a double newline after committer
+	message = append(message, fmt.Sprintf("committer %s <%s> %d %s\n", committer.GetName(), committer.GetEmail(), committer.GetDate().Unix(), committer.GetDate().Format("-0700")))
+	message = append(message, fmt.Sprintf("%s", *commit.Message))
```
`message = append(message, commit.GetMessage())`

anandkumarpatel replied:
Again no getter, so will just use Message directly

anandkumarpatel commented:
@gmlewis Fixed the issues.
* checked for empty message
* added respective test
* simplified adding message (removed sprintf)

gmlewis commented:
friendly ping @gauntface

anandkumarpatel commented:
@gauntface I would assume if `commit.Verification` is set then the user wants to use that to sign instead of the `commit.SigningKey`. 
My assumption is that a default `SigningKey` is always passed and that `Verification` would only be passed in special cases which should override `SigningKey`. 

Are my assumptions invalid?

gmlewis commented:
Good discussion. Thanks, @gauntface and @anandkumarpatel !
I like the fact that the comment for `SigningKey` says `Ignored if Verification.Signature is defined.` so there shouldn't be much confusion in this regard.
Merging.

gmlewis commented on go.mod:
```diff
@@ -3,6 +3,7 @@ module github.com/google/go-github/v26
 require (
 	github.com/golang/protobuf v1.2.0 // indirect
 	github.com/google/go-querystring v1.0.0
+	github.com/jchavannes/go-pgp v0.0.0-20170517064339-cb4b0b0ac8cf
```
Why is this line needed if we are only importing `golang.org/x/crypto/openpgp`?
Can you please revert the changes to `go.mod` and `go.sum` if these changes aren't needed?

anandkumarpatel replied:
Done! They were there with my first implementation (had nice sugar functions) but I decided to stick with `golang.org/x/crypto/openpgp` and stay away from the sugar.

gmlewis commented:
Thank you, @anandkumarpatel!
Merging.
